arwen/contexts/output.go

```diff
@@ -292,20 +292,32 @@ func (context *outputContext) Transfer(destination []byte, sender []byte, gasLim
 	return nil
 }
 
-// TransferESDT makes the asynchronous call and exports the data if it is cross shard
-func (context *outputContext) TransferESDT(destination []byte, sender []byte, tokenIdentifier []byte, value *big.Int, input []byte, gasLimit uint64) error {
+// TransferESDT makes the esdt transfer and subsequent call and exports the data if it is cross shard
+func (context *outputContext) TransferESDT(
+	destination []byte,
+	sender []byte,
+	tokenIdentifier []byte,
+	value *big.Int,
+	callInput *vmcommon.ContractCallInput,
+	gasLimit uint64,
+) error {
 	err := context.host.ExecuteESDTTransfer(destination, sender, tokenIdentifier, value)
 	if err != nil {
 		return err
 	}
 
-	if context.host.Blockchain().IsSmartContract(destination) && !context.host.AreInSameShard(sender, destination) {
-		if gasLimit > context.host.Metering().GasLeft() {
-			return arwen.ErrNotEnoughGas
-		}
+	if context.host.Blockchain().IsSmartContract(destination) && callInput != nil {
+		if !context.host.AreInSameShard(sender, destination) {
+			if gasLimit > context.host.Metering().GasLeft() {
+				return arwen.ErrNotEnoughGas
+			}
```

Comment:
Gas checking should happen irrespective of same-shard / cross-shard.

Reply:
done